CrmAuthenticable to CiviCrm

isaacmg410 opened the pull request.

Commit:
initial commit

Commit:
working on register user

Commit:
initial register applied

Commit:
rename perform_Request methods

Commit:
add devise custom locales

Commit:
modify engine

Commit:
add sessions controller actions

Commit:
working on after_confirmation

Commit:
add verification/authorization crm login

Commit:
working on it

Commit:
rename module

Commit:
missing files on rename

Commit:
apply new rest queries to civicrm

isaacmg410 commented:
@tramuntanal @aitorlb this PR is an initial state. 
Specs are missing, and some text refactors pending

tramuntanal commented on .ruby-version:
```diff
@@ -0,0 +1 @@
+2.5.3
```
better latest Ruby version: 2.6.3

tramuntanal commented on README.md:
```diff
@@ -49,7 +51,7 @@ crm_login:
 ## Versioning
 
-`Decidim::Erc::CrmLogin` depends directly on `Decidim::Core` in `0.18.0` version.
+`Decidim::Erc::CrmAuthenticable` depends directly on `Decidim::Core` in `0.18.0` version.
```
Gemfile says v0.17

Commit:
apply some refactors

tramuntanal commented on app/controllers/decidim/devise/registrations_controller.rb:
```diff
@@ -0,0 +1,85 @@
+              sign_up_as: "user",
+            }.merge(extra_params)
+          )
+          flash[:notice] =  "S'ha trobat l'usuari"
```
only catalan is ok?

tramuntanal commented on app/decorators/decidim/create_registration_decorator.rb:
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+require_dependency 'decidim/create_registration'
+Decidim::CreateRegistration.class_eval do
```
Can you please what is this decoratori overriding?
I mean, explain what it is doing different from the original command.

tramuntanal commented on app/decorators/decidim/user_decorator.rb:
```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+require_dependency 'decidim/user'
+Decidim::User.class_eval do
```
Can you please what is this decorator overriding?
I mean, explain what it is doing different from the original class.
If all are helper methods you can explain simply that.

tramuntanal reviewed (commented):
I'll start with some comments and when we resolve them i'll continue :D

Commit:
fix scopes, pending comarcal section

aitorlb committed:
Add rubocop configuration files

aitorlb committed:
Update README

aitorlb committed:
Update dependencies (master branch)

aitorlb committed:
Refactor civi_crm service

aitorlb committed:
Refactor authorization handler

aitorlb committed:
Refactor user decorator

aitorlb committed:
Refactor step 1 form

aitorlb committed:
Refactor step 2 form

aitorlb committed:
Update views

aitorlb committed:
Refactor registration command decorator

aitorlb committed:
Refactor registrations controller

aitorlb committed:
Add sessions controller

aitorlb committed:
Add tests for custom action authorizer

aitorlb committed:
Update module

aitorlb committed:
Rspec config files

aitorlb committed:
Add some missing locales

aitorlb committed:
Add registration and login system tests

aitorlb committed:
Add rake task to import "scopes" from CiviCRM

aitorlb committed:
Fix :ca locale key

aitorlb committed:
Refactor client and update rake task

aitorlb committed:
Update command decorator

aitorlb committed:
Update specs

aitorlb committed:
Update rake task and module constant

aitorlb committed:
Add migration and update rakefile

aitorlb committed:
Update README

aitorlb committed:
Change attribute reader

aitorlb committed:
Update rake task

aitorlb committed:
Add error message

aitorlb committed:
Lil refactor

aitorlb committed:
Drop scope and crm validations for admins

aitorlb committed:
Simplify extended_data form related attributes logic

aitorlb committed:
Refactor rake task

aitorlb committed:
Add comments

aitorlb committed:
Update valid_membership validation

aitorlb committed:
Remove skipped tests

aitorlb committed:
Add comments

aitorlb committed:
Rubocop all the things

aitorlb committed:
Add rake task spec

Commit:
Merge pull request #2 from CodiTramuntana/add/tests

Add: tests

aitorlb committed:
Update CrmAuthenticableAuthorizationHandler#valid_membership?

Validates the contact returned from CiviCRM is current on payment

aitorlb committed:
Require environment in rake task

aitorlb committed:
Change Ruby version

aitorlb committed:
Update tests

isaacmg410 commented on app/services/decidim/erc/crm_authenticable/crm_authenticable_action_authorizer.rb:
```diff
@@ -0,0 +1,30 @@
+        def unmatched_fields
+          @unmatched_fields ||= (valued_options_keys & authorization.metadata.to_h.keys).each_with_object({}) do |field, unmatched|
+            if field == "join_date"
+              required_value = (Date.current - options[field].to_i.months)
```
Ho poso com a nota, aquí crec que haurem de canviar el valor Date.current per la Data del plenari. Ho revisem abans de fer res.

aitorlb committed:
Use Base64#strict_encoding64 to avoid newlines being added

The problem was that document_number was being encoded as "MjYzMTgzNDRM\n" (example),
but the RegistrationForm received "MjYzMTgzNDRM\r\n". This caused the User to be saved
with the wrong String and then IdentityDocumentForm was not being able to validate that
no other User in the Organization was registered with the same encoded data.

aitorlb committed:
Update rubocop.yml

Commit:
Change: CiviCRM membership validation (#4)

Commit:
Change: Freeze registration data (#3)

Commit:
Change: Scopes Mapping (#6)

Commit:
Change: Use initials as nicknames (#5)

* Change: Use initials as nicknames

* Add model test

aitorlb committed:
Fix constant name in rake task

aitorlb committed:
Rubocop all the things

aitorlb committed:
Update test

aitorlb committed:
Update Gemfile and gemspec

aitorlb committed:
Drop unused file

aitorlb committed:
Update action authorizer

aitorlb committed:
Use IDs instead of names to filter comarcal exceptions

aitorlb committed:
Normalize locales
